File: app/src/main/java/felixwiemuth/simplereminder/data/Reminder.java
/*
 * Copyright (C) 2018-2021 Felix Wiemuth and contributors (see CONTRIBUTORS.md)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package felixwiemuth.simplereminder.data;

import androidx.annotation.NonNull;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonDeserializer;
import com.google.gson.JsonPrimitive;
import com.google.gson.JsonSerializer;
import com.google.gson.reflect.TypeToken;

import java.lang.reflect.Type;
import java.util.Calendar;
import java.util.Collection;
import java.util.Date;
import java.util.List;

import lombok.Builder;
import lombok.Getter;
import lombok.Setter;

/**
 * @author Felix Wiemuth
 */
@Getter
public class Reminder implements Comparable<Reminder> {
    private static Gson gson;

    private static final int OFFSET_REQUEST_CODE_ADD_REMINDER_DIALOG_ACTIVITY_PENDING_INTENT = 1000000;

    /**
     * Status of saved reminders.
     */
    public enum Status {
        /**
         * The reminder has been scheduled but is not due yet.
         */
        SCHEDULED,
        /**
         * The reminder is due and the notification has been sent.
         */
        NOTIFIED,
        /**
         * The reminder has been marked as "done" by the user.
         */
        DONE
    }

    /**
     * ID of the reminder, also used for notifications. Must be in the interval [0,{@link #OFFSET_REQUEST_CODE_ADD_REMINDER_DIALOG_ACTIVITY_PENDING_INTENT}) and even (used by {@link felixwiemuth.simplereminder.ReminderService for correct scheduling}).
     */
    private final int id;
    /**
     * Reminder's due date.
     */
    private @Setter Date date;
    /**
     * The interval in minutes this reminder should be repeated until dismissed.
     * This field is optional. A value <= 0 (or omitting in JSON) means that nagging is disabled, which is the default.
     * @since 0.9.9
     */
    private final int naggingRepeatInterval;
    private @Setter String text;
    private @Setter ReminderType reminderType;
    private @Setter Status status;

    public enum ReminderType {
        NORMAL,
        NAGGING,
        ALARM,
    }

    @Builder //(builderClassName = "Builder")
    public Reminder(int id, @NonNull Date date, int naggingRepeatInterval, @NonNull String text, @NonNull ReminderType reminderType) {
        if (id < 0) {
            throw new IllegalArgumentException("Id must be >= 0.");
        }
        this.id = id;
        this.date = date;
        this.naggingRepeatInterval = naggingRepeatInterval;
        this.text = text;
        this.reminderType = reminderType;
        this.status = Status.SCHEDULED;
    }

    public Calendar getCalendar() {
        Calendar c = Calendar.getInstance();
        c.setTime(date);
        return c;
    }

    @Override
    public int compareTo(Reminder reminder) {
        return date.compareTo(reminder.date);
    }

    public static String toJson(List<Reminder> reminders) {
        return getGson().toJson(reminders);
    }

    public static List<Reminder> fromJson(String json) {
        Type collectionType = new TypeToken<Collection<Reminder>>() {
        }.getType();
        return getGson().fromJson(json, collectionType);
    }

    private static Gson getGson() {
        if (gson == null) {
            gson = new GsonBuilder()
                    .registerTypeAdapter(Date.class, (JsonDeserializer<Date>) (json, typeOfT, context) -> new Date(json.getAsJsonPrimitive().getAsLong()))
                    .registerTypeAdapter(Date.class, (JsonSerializer<Date>) (date, type, jsonSerializationContext) -> new JsonPrimitive(date.getTime()))
                    .create();
        }
        return gson;
    }

    public int getRequestCodeAddReminderDialogActivityPendingIntent() {
        return getRequestCodeAddReminderDialogActivityPendingIntent(id);
    }

    public static int getRequestCodeAddReminderDialogActivityPendingIntent(int reminderID) {
        return OFFSET_REQUEST_CODE_ADD_REMINDER_DIALOG_ACTIVITY_PENDING_INTENT + reminderID;
    }

    public boolean isNagging() {
        return reminderType.equals(ReminderType.NAGGING);
    }

    public long getNaggingRepeatIntervalInMillis() {
        return 60*1000*naggingRepeatInterval;
    }
}

File: app/src/main/java/felixwiemuth/simplereminder/data/Reminder.kt
/*
 * Copyright (C) 2018-2022 Felix Wiemuth and contributors (see CONTRIBUTORS.md)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package felixwiemuth.simplereminder.data

import felixwiemuth.simplereminder.data.Reminder.Companion.MAX_REMINDER_ID
import felixwiemuth.simplereminder.util.DateSerializer
import kotlinx.serialization.Serializable
import kotlinx.serialization.builtins.ListSerializer
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import java.util.*

@Serializable
data class Reminder
constructor(
    /**
     * ID of the reminder, also used for notifications. Must be in the range 0..[MAX_REMINDER_ID] and even.
     */
    val id: Int,

    /**
     * Reminder's due date.
     */
    @Serializable(with = DateSerializer::class)
    val date: Date,

    /**
     * The interval in minutes with which this reminder should be repeated until dismissed.
     * This field is optional. A value <= 0 (or omitting in JSON) means that nagging is disabled, which is the default.
     * @since 0.9.9
     */
    val naggingRepeatInterval: Int = 0,

    val text: String = "",

    var status: Status = Status.SCHEDULED
) : Comparable<Reminder> {
    /**
     * Status of saved reminders.
     */
    enum class Status {
        /**
         * The reminder has been scheduled but is not due yet.
         */
        SCHEDULED,

        /**
         * The reminder is due and the notification has been sent.
         */
        NOTIFIED,

        /**
         * The reminder has been marked as "done" by the user.
         */
        DONE
    }

    init {
        require(id in 0..MAX_REMINDER_ID && id % 2 == 0) { "Id must be even, >= 0 and <= $MAX_REMINDER_ID." }
    }

    /**
     * Get a new [Calendar] instance set to the reminder's date.
     */
    val calendar: Calendar
        get() {
            val c = Calendar.getInstance()
            c.time = date
            return c
        }

    override fun compareTo(other: Reminder): Int {
        return date.compareTo(other.date)
    }

    val isNagging: Boolean
        get() = naggingRepeatInterval > 0
    val naggingRepeatIntervalInMillis: Long
        get() = (60 * 1000 * naggingRepeatInterval).toLong()

    companion object {
        const val MAX_REMINDER_ID = 1000000

        @JvmStatic
        fun builder(date: Date, text: String): Builder = Builder(date = date, text = text)

        @JvmStatic
        fun toJson(reminders: List<Reminder?>?): String =
            Json.encodeToString(reminders)

        @JvmStatic
        fun fromJson(json: String): List<Reminder> =
            Json.decodeFromString(ListSerializer(serializer()), json)
    }

    /**
     * Used to construct a [Reminder] step by step. This allows the ID for a new reminder to be first created when adding the reminder to the storage.
     */
    data class Builder
    @JvmOverloads
    constructor(
        @JvmField
        var id: Int? = null,
        @JvmField
        val date: Date,
        @JvmField
        var naggingRepeatInterval: Int = 0,
        @JvmField
        val text: String = "",
        @JvmField
        var status: Status = Status.SCHEDULED
    ) {
        fun build() = Reminder(requireNotNull(id), date, naggingRepeatInterval, text, status)
    }
}